pkg/storaged/btrfs/volume.jsx

```diff
@@ -0,0 +1,190 @@
+import { DescriptionList } from "@patternfly/react-core/dist/esm/components/DescriptionList/index.js";
+import { Stack, StackItem } from "@patternfly/react-core/dist/esm/layouts/Stack/index.js";
+import { VolumeIcon } from "../icons/gnome-icons.jsx";
+
+import { new_card, new_page, get_crossrefs, ChildrenTable, PageTable, StorageCard, StorageDescription } from "../pages.jsx";
+import { StorageUsageBar, StorageLink } from "../storage-controls.jsx";
+import { fmt_size_long, validate_fsys_label } from "../utils.js";
+import { btrfs_usage, btrfs_is_volume_mounted } from "./utils.jsx";
+import { dialog_open, TextInput } from "../dialog.jsx";
+import { make_btrfs_subvolume_page } from "./subvolume.jsx";
+import { mounting_dialog } from "../filesystem/mounting-dialog.jsx";
+
+const _ = cockpit.gettext;
+
+function volume_mount(volume) {
+    const block = client.blocks[volume.path];
+    mounting_dialog(client, block, "mount", null);
+}
+
+/*
+ * Udisks is a disk/block library so it manages that, btrfs turns this a bit
+ * around and has one "volume" which can have multiple blocks by a unique uuid.
+ *
+ * Cockpit shows Btrfs as following:
+ *
+ * -> btrfs subvolume
+ *    -> btrfs volume
+ *        -> btrfs device
+ *            -> block device
+ */
+export function make_btrfs_volume_card(parent, uuid) {
+    const block_devices = client.uuids_btrfs_blocks[uuid];
+    const block_btrfs = client.blocks_fsys_btrfs[block_devices[0].path];
+    const volume = client.uuids_btrfs_volume[uuid];
+    const use = btrfs_usage(client, volume);
+
+    const btrfs_disks_card = new_card({
+        title: _("btrfs devices"),
+        next: null,
+        component: BtrfsDevicesCard,
+        props: { uuid },
+    });
+
+    const name = block_btrfs.data.label || uuid;
+    const btrfs_volume_card = new_card({
+        title: _("btrfs volume"),
+        next: btrfs_disks_card,
+        page_location: ["btrfs-volume", uuid],
+        page_name: name,
+        page_icon: VolumeIcon,
+        page_size: use[1],
+        component: BtrfsVolumePage,
+        props: { block_devices, uuid, use },
+    });
+
+    const actions = [];
+    const is_mounted = btrfs_is_volume_mounted(client, block_devices);
+    if (!is_mounted) {
+        actions.push({
+            title: _("Mount"),
+            action: () => volume_mount(volume),
+        });
+    }
+
+    const subvolumes_card = new_card({
+        title: _("btrfs subvolumes"),
+        next: btrfs_volume_card,
+        component: BtrfsSubVolumesCard,
+        props: { volume },
+        // TODO: create subvolume
+        actions,
+    });
+
+    const subvolumes_page = new_page(parent, subvolumes_card);
+    make_btrfs_subvolume_pages(subvolumes_page, volume);
+}
+
+const BtrfsVolumePage = ({ card, block_devices, uuid, use }) => {
+    const block_btrfs = client.blocks_fsys_btrfs[block_devices[0].path];
+    const label = block_btrfs.data.label || "-";
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20231205-110900-7b3be2c9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)